### DISCUSSION TITLE
Reward Function Templates

### DISCUSSION THREAD

**Discussion by fionnachan** [PathmindAI/pathmind-webapp#3415] [Ideas]
@johnnyL7 requested a feature to make the reuse of reward functions more convenient when the user wants to reuse a set of previous reward functions for a new model.

The following is a sketch of my idea.

User can save any existing reward function of a non-draft Experiment as a reward function template by clicking the "Save" button on the reward function panel 
![Experiment View - Save button](https://user-images.githubusercontent.com/13184582/128683600-288ad1b3-d8e3-4011-914b-b20a2cfea9c0.png)
![Experiment View - Save Reward Function as Template popup](https://user-images.githubusercontent.com/13184582/128683753-81011b2c-87a5-4980-a39f-fa19d1930973.png)

On the Project/Model view, there is an option to create multiple experiments.
![Project Model View - New Experiment Split Button](https://user-images.githubusercontent.com/13184582/128683895-def9d72a-d5dd-4d3b-84c2-a6965ed26b1d.png)
![Project Model View - New Experiment Split Button Dropdown](https://user-images.githubusercontent.com/13184582/128683912-9d6a37db-4bb6-4f96-a855-77e2e5d8194a.png)
![Project Model View - Create New Experiments popup](https://user-images.githubusercontent.com/13184582/128683928-c9e67cf4-a041-4cf2-aabc-714d7d940faf.png)

Similarly, at the final step of the Model Upload process, the user can choose to create multiple experiments. Clicking that button would bring the user to a page which asks for the same information as demonstrated above.
![Upload Model View - Create Experiment buttons](https://user-images.githubusercontent.com/13184582/128684126-be2f5a85-1467-4a8c-807c-9809ac27e871.png)

Add "Saved" Page to Main Menu
![Menu - Saved](https://user-images.githubusercontent.com/13184582/128683546-32e65c7f-dfb1-40d7-8cbe-49be09a5a6a1.png)

User viewing their saved reward function templates. Clicking the trash button would show a confirmation dialog asking if the user wants to delete the reward function template. It should delete the reward function template from the DB.
![Saved - Reward Function Templates list](https://user-images.githubusercontent.com/13184582/128684224-011fbcc9-01a0-49ea-98ba-a3f212c04f4d.png)

User can edit their saved reward function templates
![Saved - Edit Reward Function Template](https://user-images.githubusercontent.com/13184582/128684489-d6f50640-ce31-438a-8cd3-af0b62f3ad89.png)

**Comment by slinlee** [comment]
I like this. There's a lot of good thinking, @fionnachan 

Here are some questions in no particular order:
- What if a user wants to start with a template and then modify it (maybe to match a change in a reward var name?)
  - If starting multiple experiments, maybe they're each experiment is created as a draft until the user clicks 'start training'?
  - On the new experiment page, we could add a dropdown to select from the templates. This would populate the reward function editing area. This lets the user preview or edit the reward the function before starting training.
- When starting multiple experiments, how about we just show a list of all the templates and the user can check a checkbox to opt-in each reward function they want to start. This way they don't have to think of the number of experiments first
- Do we need a 'favorite' template function in addition to the ability to save templates? This would be to differentiate the current best choices.

**Reply by fionnachan** [reply]
1) Yea I was thinking that we create multiple draft experiments for them and not run them immediately so that should solve the problem?
    - Yes that's great :D
=> I'll add the new experiment view mockup in the comments below later

2) Oh ok. Do you think we should show the reward function code itself too or just the reward function template names?
=> I'll create a mockup for this too

3) That sounds good but I'd prefer it to be a separate "feature enhancement" ticket after this "reward function template" work

Btw would we want to give free & paid users a different total limit of reward function templates they can have?

**Reply by slinlee** [reply]
> paid vs free

how about we make templates a paid only feature? Sound like a good idea @ejunprung ?

**Reply by ejunprung** [reply]
Yeah, definitely. I think we everything we do going forward (except fixing bugs, sales/marketing stuff, etc.) should be paid only.

**Comment by fionnachan** [comment]
New Experiment View
![New Experiment View - Reward Function Editor toolbar](https://user-images.githubusercontent.com/13184582/128804271-88d6dccc-f5e8-4f32-a0b1-490e6694bf7b.png)
![New Experiment View - Choose a Reward Function Template popup](https://user-images.githubusercontent.com/13184582/128804278-6386be54-0eec-41c0-a4d0-a7f97c643aac.png)

List box
The only caveat here is that the list could be long enough that the list box will need a scrollbar
![Project Model View - Create New Experiments popup with listbox](https://user-images.githubusercontent.com/13184582/128804299-dcba2d47-c3b5-437b-b3f9-41b862b41866.png)

**Comment by ejunprung** [comment]
Are reward templates specific to particular model versions? What happens if the user changes the reward syntax and breaks the template? I think 90%+ of the time, the rewards are not portable.

**Reply by slinlee** [reply]
This is meant exactly to port reward functions between model revisions. The way that we're trying to get around that is to create the new experiments as 'drafts' first. Then the user can edit them and also see which variables are invalid.

**Comment by slinlee** [comment]
> Add "Saved" Page to Main Menu
> ![Menu - Saved](https://user-images.githubusercontent.com/13184582/128683546-32e65c7f-dfb1-40d7-8cbe-49be09a5a6a1.png)
> 
> User viewing their saved reward function templates. Clicking the trash button would show a confirmation dialog asking if the user wants to delete the reward function template. It should delete the reward function template from the DB.
> ![Saved - Reward Function Templates list](https://user-images.githubusercontent.com/13184582/128684224-011fbcc9-01a0-49ea-98ba-a3f212c04f4d.png)

@fionnachan I think I missed this section when I first read through. To me, the scope of the templates is per project, since that's what would have the same objectives and goals across models. 

So a user wouldn't use a coffeeshop function on the moonlander, since the coffeeshop would be measuring things like wait time and customers served.

The different saved functions would be changes like, "double the importance of fuel savings" or "factor in customer satisfaction" for a single project.

**Reply by fionnachan** [reply]
I think it's easier to organise the templates on one dedicated page. If I am the user, I would want to know how many templates I have at any give time, and to be able to edit / remove them.

We can move the "Templates" button to the Project/Model view instead if a standalone section is not a good option and only show the templates for that project
![image](https://user-images.githubusercontent.com/13184582/129119720-a3f53dc7-25ec-4af0-b3f8-9e106adfbbcd.png)

Originally I thought it's up to the user to keep which templates they have and show all of them on the dropdown / list. I agree that showing only those for the project would be better.

**Reply by slinlee** [reply]
Yep the same list of templates but just on the project level sounds good. For a minimal first pass I think the feature would still work without the list of templates too.
